File: pdu/csharp_v2/sensor_msgs/pdu_conv_MagneticField.cs
using System;
using System.Collections.Generic;

using Hakoniwa.Pdu.CSharpV2;
using Hakoniwa.Pdu.CSharpV2.sensor_msgs;
using Hakoniwa.Pdu.CSharpV2.std_msgs;
using Hakoniwa.Pdu.CSharpV2.geometry_msgs;

namespace Hakoniwa.Pdu.CSharpV2.sensor_msgs
{
    public static class MagneticFieldConverter
    {
        public static MagneticField PduToMsg(byte[] binaryData)
        {
            var obj = new MagneticField();
            var meta = PduMetaData.Parse(binaryData);
            BinaryReadRecursive(meta, binaryData, obj, PduMetaData.PduMetaDataSize);
            return obj;
        }

        public static byte[] MsgToPdu(MagneticField obj)
        {
            var baseAllocator = new DynamicAllocator();
            var writer = new BinaryWriterContainer();
            BinaryWriteRecursive(0, writer, baseAllocator, obj);
            return PduRuntime.BuildPdu(baseAllocator, writer);
        }

        public static void BinaryReadRecursive(PduMetaData meta, byte[] binaryData, MagneticField obj, int baseOff)
        {
            obj.header = new Header();
            HeaderConverter.BinaryReadRecursive(meta, binaryData, obj.header, baseOff + 0);
            obj.magnetic_field = new Vector3();
            Vector3Converter.BinaryReadRecursive(meta, binaryData, obj.magnetic_field, baseOff + 136);
            obj.magnetic_field_covariance = new List<double>();
            for (var i = 0; i < 9; i++) {
                obj.magnetic_field_covariance.Add(PduRuntime.ReadFloat64(binaryData, baseOff + 160 + (i * 8)));
            }
        }

        public static void BinaryWriteRecursive(int parentOff, BinaryWriterContainer writer, DynamicAllocator allocator, MagneticField obj)
        {
            HeaderConverter.BinaryWriteRecursive(parentOff + 0, writer, allocator, obj.header);
            Vector3Converter.BinaryWriteRecursive(parentOff + 136, writer, allocator, obj.magnetic_field);
            for (var i = 0; i < obj.magnetic_field_covariance.Count; i++) {
                allocator.Add(PduRuntime.GetBinaryForFloat64(obj.magnetic_field_covariance[i]), parentOff + 160 + (i * 8));
            }
        }
    }
}

File: pdu/javascript/sensor_msgs/pdu_conv_MagneticField.js
import * as PduUtils from '../pdu_utils.js';
import { MagneticField } from './pdu_jstype_MagneticField.js';
import { Header } from '../std_msgs/pdu_jstype_Header.js';
import { binary_read_recursive_Header, binary_write_recursive_Header } from '../std_msgs/pdu_conv_Header.js';
import { Vector3 } from '../geometry_msgs/pdu_jstype_Vector3.js';
import { binary_read_recursive_Vector3, binary_write_recursive_Vector3 } from '../geometry_msgs/pdu_conv_Vector3.js';


/**
 * Deserializes a binary PDU into a MagneticField object.
 * @param {ArrayBuffer} binary_data
 * @returns { MagneticField }
 */
export function pduToJs_MagneticField(binary_data) {
    const js_obj = new MagneticField();
    const meta_parser = new PduUtils.PduMetaDataParser();
    const meta = meta_parser.load_pdu_meta(binary_data);
    if (meta === null) {
        throw new Error("Invalid PDU binary data: MetaData not found or corrupted");
    }
    binary_read_recursive_MagneticField(meta, binary_data, js_obj, meta.base_off);
    return js_obj;
}

export function binary_read_recursive_MagneticField(meta, binary_data, js_obj, base_off) {
    const view = new DataView(binary_data);
    const littleEndian = true;
    // member: header, type: std_msgs/Header (struct)

    {
        const tmp_obj = new Header();
        binary_read_recursive_Header(meta, binary_data, tmp_obj, base_off + 0);
        js_obj.header = tmp_obj;
    }
    
    // member: magnetic_field, type: geometry_msgs/Vector3 (struct)

    {
        const tmp_obj = new Vector3();
        binary_read_recursive_Vector3(meta, binary_data, tmp_obj, base_off + 136);
        js_obj.magnetic_field = tmp_obj;
    }
    
    // member: magnetic_field_covariance, type: float64 (primitive)

    
    {
        const array_bin = PduUtils.readBinary(binary_data, base_off + 160, 72);
        js_obj.magnetic_field_covariance = PduUtils.binToArrayValues("float64", array_bin, 9, 72 / 9);
    }
    
    return js_obj;
}

/**
 * Serializes a MagneticField object into a binary PDU.
 * @param { MagneticField } js_obj
 * @returns {ArrayBuffer}
 */
export function jsToPdu_MagneticField(js_obj) {
    const base_allocator = new PduUtils.DynamicAllocator();
    const bw_container = new PduUtils.BinaryWriterContainer(new PduUtils.PduMetaData());

    binary_write_recursive_MagneticField(0, bw_container, base_allocator, js_obj);

    const base_data_size = base_allocator.size();
    const heap_data_size = bw_container.heap_allocator.size();
    
    bw_container.meta.heap_off = PduUtils.PDU_META_DATA_SIZE + base_data_size;
    bw_container.meta.total_size = bw_container.meta.heap_off + heap_data_size;

    const final_buffer = new ArrayBuffer(bw_container.meta.total_size);
    const final_view = new Uint8Array(final_buffer);

    PduUtils.writeBinary(final_view, 0, bw_container.meta.to_bytes());
    PduUtils.writeBinary(final_view, bw_container.meta.base_off, base_allocator.toArray());
    PduUtils.writeBinary(final_view, bw_container.meta.heap_off, bw_container.heap_allocator.toArray());

    return final_buffer;
}

export function binary_write_recursive_MagneticField(parent_off, bw_container, allocator, js_obj) {
    const littleEndian = true;
    // member: header, type: std_msgs/Header (struct)

    {
        binary_write_recursive_Header(parent_off + 0, bw_container, allocator, js_obj.header);
    }
    
    // member: magnetic_field, type: geometry_msgs/Vector3 (struct)

    {
        binary_write_recursive_Vector3(parent_off + 136, bw_container, allocator, js_obj.magnetic_field);
    }
    
    // member: magnetic_field_covariance, type: float64 (primitive)

    
    {
        const buffer = PduUtils.typesToBin("float64", js_obj.magnetic_field_covariance, 72 / 9);
        allocator.add(buffer, parent_off + 160);
    }
    
}
